iina/JavasctiptDevTool.swift

```diff
@@ -0,0 +1,506 @@
+          proxy.scrollTo(id, anchor: .bottom)
+        }
+      }
+      VStack(alignment: .leading, spacing: 0) {
+        // Toolbar
+        HStack {
+          // Run code
+          Button(action: { executePrompt() }) {
+            Image(systemName: "play.circle")
+              .renderingMode(.template)
+              .foregroundColor(.accentColor)
+          }
+          .help("Run code")
+          .buttonStyle(PlainButtonStyle())
+          .keyboardShortcut(.return, modifiers: .command)
+          // Show global objects
+          Button(action: { executePrompt(printGlobalObject: true) }) {
+            Image(systemName: "point.3.filled.connected.trianglepath.dotted")
+              .renderingMode(.template)
+              .foregroundColor(.secondary)
+          }
+          .buttonStyle(PlainButtonStyle())
+          .help("Print global objects")
+          Spacer()
+          Text("Cmd+Return to run code")
+            .font(.footnote)
+            .foregroundColor(.secondary)
+        }
+        .padding([.top, .horizontal], 8)
+        // Editor
+        CommandEditor(text: $input)
+          .frame(maxHeight: 100)
+          .padding(.all, 8)
+
+      }
+    }
+    .listStyle(.plain)
+  }
+
+  private func executePrompt(printGlobalObject: Bool = false) {
+    let source = printGlobalObject ? "$global" : input.trimmingCharacters(in: .whitespacesAndNewlines)
+    if source.isEmpty {
+      return
+    }
+    events.addPrompt(source)
+    events.postChange()
+
+    let jsResult: JSEvent.Result
+
+    let result = source == "$global" ?
+    jsContext.globalObject :
+    jsContext.evaluateScript(source)
+
+    if let result = result {
+      if result.isNumber {
+        jsResult = .number(result.toNumber())
+      } else if result.isString {
+        jsResult = .string(result.toString())
+      } else if result.isBoolean {
+        jsResult = .boolean(result.toBool())
+      } else if result.isNull {
+        jsResult = .null
+      } else if result.isUndefined {
+        jsResult = .undefined
+      } else if result.isArray {
+        let array = result.toArray()!
+        jsResult = .array(
+          "Array (\(array.count))", array.enumerated().map { ("\($0.offset)", "\($0.element)") } )
+      } else {
+        let object = result.toObject()!
+        if let dict = object as? [String: Any] {
+          jsResult = .object(
+            result.toString() ?? "", dict.keys.map { ($0, "\(dict[$0] ?? "nil")") })
+        } else {
+          jsResult = .opaqueObject(result.toString())
+        }
+      }
+      events.addResult(jsResult)
+      events.postChange()
+    }
+
+    if !printGlobalObject {
+      input = ""
+    }
+  }
+}
+
+@available(macOS 12.0, *)
+struct JavasctiptDevTool_Previews: PreviewProvider {
+  static var previews: some View {
+    JavasctiptDevTool(jsContext: JSContext())
+  }
+}
+
+
+// MARK: - Views
+
+@available(macOS 12.0, *)
+struct CommandEditor: NSViewRepresentable {
```

Comment:
Unsure how to represent this but I generally find it useful to know why something is not implemented in SwiftUI, so that it can be updated later if things change. Is it because text selection is not available in `TextField`?

Reply:
Not sizable, shoddy multiline support, unable to capture key events, etc.